### PR TITLE
Switch to k8s client-go

### PR DESCRIPTION
**Pull request by daxmc99**
Fixes https://github.com/sourcegraph/sourcegraph/issues/11804

This is needed as part of https://github.com/sourcegraph/sourcegraph/issues/13878 since without this we cannot use a current client-go (this client is ~2 years old)

### CODE REVIEW

File: cmd/frontend/internal/app/debugproxies/scanner.go [hunk 1]
```diff
@@ -198,10 +180,16 @@ func (cs *clusterScanner) scanCluster() {
...
 }
 
-// fromStrPtr returns *s. If s is nil the empty string is returned.
-func fromStrPtr(s *string) string {
-	if s == nil {
-		return ""
+// namespace returns the namespace the pod is currently running in
+// this is done because the k8s client we previously used set the namespace
+// when the client was created, the official k8s client does not
+func namespace() string {
```
Comment (uwedeportivo) [inline cmd/frontend/internal/app/debugproxies/scanner.go:186]: can you please rewrite this a little to log the error if it happens.

**Review by uwedeportivo — APPROVED** [review]
LGTM, please add keegan or somebody from cloud as a reviewer. endpoint.go is used to discover zoekt service endpoints so it's rather important :-)

File: cmd/frontend/internal/app/debugproxies/scanner.go (hunk 1 above)
Comment (daxmc99) [reply cmd/frontend/internal/app/debugproxies/scanner.go:186]: Rewritten

**Review by bobheadxi — APPROVED** [review]

**Review by uwedeportivo — APPROVED** [review]
LGTM

**Comment by keegancsmith** [comment]
I'll review the code soon, but at a high level this pulls in a crap load of transitive dependencies. We used to use client-go, and constantly ran into dependency issues due to how much it depends on and how they change.

Why do you need client-go? Does this simple client not solve it for you?

File: cmd/frontend/internal/app/debugproxies/scanner.go [hunk 2]
```diff
@@ -102,21 +85,23 @@ func (cs *clusterScanner) runEventLoop() {
 // watchEndpointEvents uses the k8s watch API operation to watch for endpoint events. Spins forever unless an error
 // occurs that would necessitate creating a new watcher. The caller will then call again creating the new watcher.
 func (cs *clusterScanner) watchEndpointEvents() (bool, error) {
-	watcher, err := cs.client.Watch(context.Background(), cs.client.Namespace(), new(corev1.Endpoints))
+
+	// TODO(Dax): Rewrite this to used NewSharedInformerFactory from k8s/client-go
```
Comment (keegancsmith) [inline cmd/frontend/internal/app/debugproxies/scanner.go:89]: you may have to look at older versions of this repo, but I believe we used to use informer back when this package used client-go.

File: cmd/frontend/internal/app/debugproxies/scanner.go [hunk 3]
```diff
@@ -102,21 +85,23 @@ func (cs *clusterScanner) runEventLoop() {
...
 	if err != nil {
 		return false, fmt.Errorf("k8s client.Watch error: %w", err)
 	}
-	defer watcher.Close()
+	defer watcher.Stop()
 
 	for {
-		var eps corev1.Endpoints
-		eventType, err := watcher.Next(&eps)
+		event := <-watcher.ResultChan()
```
Comment (keegancsmith) [inline cmd/frontend/internal/app/debugproxies/scanner.go:98]: you should be able to do `for event := range watcher.ResultChan()`

File: cmd/frontend/internal/app/debugproxies/scanner.go [hunk 4]
```diff
@@ -198,10 +180,21 @@ func (cs *clusterScanner) scanCluster() {
...
-	if s == nil {
-		return ""
+// namespace returns the namespace the pod is currently running in
+// this is done because the k8s client we previously used set the namespace
+// when the client was created, the official k8s client does not
+func namespace() string {
+	const filename = "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
+	data, err := ioutil.ReadFile(filename)
+	if err != nil {
+		log15.Warn("unable to read ", filename, " using \"default\" ns")
```
Comment (keegancsmith) [inline cmd/frontend/internal/app/debugproxies/scanner.go:190]: this isn't how log15 works, it is a structured logger.

```suggestion
		log15.Warn("endpoint: falling back to kubernetes default namespace", "filename", filename, "error", err)
```

File: cmd/frontend/internal/app/debugproxies/scanner.go [hunk 5]
```diff
@@ -198,10 +180,21 @@ func (cs *clusterScanner) scanCluster() {
...
+		return "default"
+	}
+
+	ns := strings.TrimSpace(string(data))
+	if len(ns) == 0 {
```
Comment (keegancsmith) [inline cmd/frontend/internal/app/debugproxies/scanner.go:195]: minor: idiomatic to check against empty string
```suggestion
	if ns == "" {
```

File: cmd/frontend/internal/app/debugproxies/scanner.go [hunk 6]
```diff
@@ -198,10 +180,21 @@ func (cs *clusterScanner) scanCluster() {
...
+		log15.Warn("file: ", filename, " empty using \"default\" ns")
```
Comment (keegancsmith) [inline cmd/frontend/internal/app/debugproxies/scanner.go:196]:
```suggestion
		log15.Warn("endpoint: falling back to kubernetes default namespace", "error", filename + " is empty")
```

File: internal/endpoint/endpoint.go [hunk 7]
```diff
@@ -172,40 +174,47 @@ func (m *Map) getUrls() (*hashMap, error) {
...
 
-func inform(client *k8s.Client, m *Map, u *k8sURL) error {
-	watcher, err := client.Watch(context.Background(), client.Namespace, new(corev1.Endpoints), k8s.QueryParam("fieldSelector", "metadata.name="+u.Service))
-	if err != nil {
-		return errors.Wrap(err, "client.Watch")
-	}
-	defer watcher.Close()
+func inform(client v1.EndpointsInterface, m *Map, u *k8sURL) error {
+
+	// TODO(Dax): We shouldn't use watch directly, use an informer here
```
Comment (keegancsmith) [inline internal/endpoint/endpoint.go:179]: FYI if watch is simpler, we should just use that. I think Informer may be simple, I can't remember exactly :) But Informer is designed to help scale / simplify access, but we only care about simplicity not scale in this case.

File: internal/endpoint/endpoint.go [hunk 8]
```diff
@@ -172,40 +174,47 @@ func (m *Map) getUrls() (*hashMap, error) {
...
+	})
+
+	defer watcher.Stop()
 
 	for {
-		var endpoints corev1.Endpoints
-		eventType, err := watcher.Next(&endpoints)
-		if err != nil {
-			return errors.Wrap(err, "watcher.Next")
+		event := <-watcher.ResultChan()
```
Comment (keegancsmith) [inline internal/endpoint/endpoint.go:187]: same as the previous code snippet, you can for loop over the chan

File: internal/endpoint/endpoint.go [hunk 9]
```diff
@@ -172,40 +174,47 @@ func (m *Map) getUrls() (*hashMap, error) {
...
+		e := event.Object
+		endpoints, ok := e.(*corev1.Endpoints)
+		if !ok {
+			return errors.Wrap(err, "object from watcher is not an endpoint")
```
Comment (keegancsmith) [inline internal/endpoint/endpoint.go:191]: have you manually tested this? Should be fine, but this sort of runtime reflection can slip through the cracks.

File: internal/endpoint/endpoint.go [hunk 10]
```diff
@@ -268,25 +277,55 @@ func newConsistentHashMap(keys []string) *hashMap {
...
+		return "default"
+	}
+
+	ns := strings.TrimSpace(string(data))
+	if len(ns) == 0 {
+		log15.Warn("file: ", filename, " empty using \"default\" ns")
+		return "default"
+	}
+	return ns
+}
```
Comment (keegancsmith) [inline internal/endpoint/endpoint.go:297]: same comments from the other place this was copied from.

**Review by keegancsmith — APPROVED** [review]
Approving. Some comments inline. This looks like it works, my main concern is that client-go is a seven tonne gorilla that I'd rather avoid in our dependency tree.

**Comment by daxmc99** [comment]
After your comment about this pulling in more dependencies, I checked the frontend binary size before and after this change:
before 68M 
after 117M 
This is a 72% increase in our binary size which feels excessive for what is essentially a refactor to use a more well-documented client. 

I knew that client-go might have some issues with other dependencies but there are no conflicts and the post 1.17.0 versions are fairly easy to update.

**Comment by daxmc99** [comment]
I also spent some time today trying to upgrade https://github.com/ericchiang/k8s but the protobuf  generation no longer works

**Comment by keegancsmith** [comment]
> After your comment about this pulling in more dependencies, I checked the frontend binary size before and after this change:
> 
> ```
> before 68M 
> after 117M
> ```
> 
> This is a 72% increase in our binary size which feels excessive for what is essentially a refactor to use a more well-documented client.



> I knew that client-go might have some issues with other dependencies but there are no conflicts and the post 1.17.0 versions are fairly easy to update.

Yeah I am guessing this has calmed down vs a few years ago.

> I also spent some time today trying to upgrade https://github.com/ericchiang/k8s but the protobuf generation no longer works

@daxmc99 you wanna pair tomorrow to try and sort this out? I'm sure making our current dep work on modern protobuf will make others happy.

**Comment by daxmc99** [comment]
The cluster QA tests are currently running via a pure bash implementation. 
The results of our pairing session are available here: 
https://github.com/ericchiang/k8s/pull/128

The upstream project has been archived though so we should avoid new uses of `ericchiang/k8s` where possible.

**Comment by keegancsmith** [comment]
> The cluster QA tests are currently running via a pure bash 
> implementation.

Cool. So just using kubectl?

> The upstream project has been archived though so we should avoid 
> new uses of `ericchiang/k8s` where possible.

Should we consider forking and maintaining it? It seems like not 
that much effort to be honest and in our interests to keep up to 
date. I imagine keeping it up to date will be easier than updating 
deps on the offical k8s client :)

**Comment by daxmc99** [comment]
> Cool. So just using kubectl?

Correct

> Should we consider forking and maintaining it?
I still hesitate to use a non-official client. I wonder if the way forward
here is actually to reduce scope further and drop a client altogether and
build our own basic client (
https://github.com/cilium/cilium/blob/master/pkg/k8s/client.go#L93:6). We
intentionally use the k8s API for very little and our current
dependency actually blocks us from using newer versions of protobuf.

**Comment by keegancsmith** [comment]
>> Should we consider forking and maintaining it?
 
> I still hesitate to use a non-official client. I wonder if the 
> way forward here is actually to reduce scope further and drop a 
> client altogether and build our own basic client

I'm all for doing that. Our use of the API is super tiny. Its 
watch and get on endpoints... and I think maybe a get on 
svc. Thats it. In fact I learnt how to do what I wanted by running 
kubectl in verbose mode anyways, which just output the urls to 
hit. Sounds like fun :)

**Comment by arussellsaw** [comment]
👋  i'm gonna pick this up to resolve conflicts & merge, as it's blocking some work for secret encryption. The GCloud KMS library depends on a newer protobuf version, and won't compile. I've pulled these changes into my local branch and everything looks fine. context: https://sourcegraph.slack.com/archives/CHPC7UX16/p1613123508259400

**Pull request merged**